diff --git a/drivers/acpi/ec.c b/drivers/acpi/ec.c
index d5309fd..0a92f7c 100644
--- a/drivers/acpi/ec.c
+++ b/drivers/acpi/ec.c
@@ -53,6 +53,7 @@
 /* EC status register */
 #define ACPI_EC_FLAG_OBF       0x01    /* Output buffer full */
 #define ACPI_EC_FLAG_IBF       0x02    /* Input buffer full */
+#define ACPI_EC_FLAG_CMD       0x08    /* cmd in data register */
 #define ACPI_EC_FLAG_BURST     0x10    /* burst mode */
 #define ACPI_EC_FLAG_SCI       0x20    /* EC-SCI occurred */
 
@@ -165,9 +166,20 @@ static int ec_transaction_done(struct acpi_ec *ec)
 
 static void start_transaction(struct acpi_ec *ec)
 {
+       int i = 0;
+
        ec->curr->irq_count = ec->curr->wi = ec->curr->ri = 0;
        ec->curr->done = false;
+
+retry:
        acpi_ec_write_cmd(ec, ec->curr->command);
+       if(!(acpi_ec_read_status(ec) & ACPI_EC_FLAG_CMD)) {
+               if(i<3) {
+                       i++;
+                       goto retry;
+               } else
+                       pr_warn("bxuflag write error\n");
+       }
 }
 
 static void advance_transaction(struct acpi_ec *ec, u8 status)
