--- coregrind/m_syswrap/syscall-arm64-linux.S.orig
+++ coregrind/m_syswrap/syscall-arm64-linux.S
@@ -162,6 +162,7 @@
 .globl ML_(blksys_complete)
 .globl ML_(blksys_committed)
 .globl ML_(blksys_finished)
+.align 8
 ML_(blksys_setup):      .quad 1b
 ML_(blksys_restart):    .quad 2b
 ML_(blksys_complete):   .quad 3b
--- VEX/priv/guest_arm64_toIR.c.orig
+++ VEX/priv/guest_arm64_toIR.c
@@ -6779,7 +6779,8 @@
    }
 
    /* -------------------- NOP -------------------- */
-   if (INSN(31,0) == 0xD503201F) {
+   if (INSN(31,0) == 0xD503201F
+       || INSN(31,0) == 0xD503203F) {
       DIP("nop\n");
       return True;
    }
