--- coregrind/m_syswrap/syscall-arm64-linux.S.orig
+++ coregrind/m_syswrap/syscall-arm64-linux.S
@@ -162,6 +162,7 @@
 .globl ML_(blksys_complete)
 .globl ML_(blksys_committed)
 .globl ML_(blksys_finished)
+.align 8
 ML_(blksys_setup):      .quad 1b
 ML_(blksys_restart):    .quad 2b
 ML_(blksys_complete):   .quad 3b
--- VEX/priv/guest_arm64_toIR.c.orig
+++ VEX/priv/guest_arm64_toIR.c
@@ -5552,6 +5552,11 @@
       DIP("dsb ish\n");
       return True;
    }
+   if (INSN(31,0) == 0xD5033F9F) {
+      stmt(IRStmt_MBE(Imbe_Fence));
+      DIP("dsb sy\n");
+      return True;
+   }
 
    /* -------------------- NOP -------------------- */
    if (INSN(31,0) == 0xD503201F) {
